test/test_document.py

```diff
@@ -396,5 +397,78 @@ def test_add_attachment(self):
         self.assertTrue(test_attach.compare(doc.attachments[0]))
 
 
+class TopLevelExtension(sbol2.TopLevel):
+
+    RDF_TYPE = 'http://example.org/test#TopLevelExtension'
+
+    def __init__(self, uri='example'):
+        super().__init__(uri=uri,
+                         type_uri=TopLevelExtension.RDF_TYPE)
+
+
+class NonTopLevelExtension(sbol2.Identified):
+
+    RDF_TYPE = 'http://example.org/test#NonTopLevelExtension'
+
+    def __init__(self, uri='example'):
+        super().__init__(uri=uri,
+                         type_uri=NonTopLevelExtension.RDF_TYPE)
+
+
+class TestDocumentExtensionObjects(unittest.TestCase):
+
+    def test_get_extension_object(self):
+        doc = sbol2.Document(CRISPR_LOCATION)
+
+        # This returns an object. A TopLevel, not sure what type it is beyond that
+        uri = 'http://sbols.org/CRISPR_Example/mKate_gene/1.0.0'
+        obj = doc.getExtensionObject(uri)
+        self.assertIsNotNone(obj)
+        self.assertIsInstance(obj, sbol2.TopLevel)
+
+        # This should raise a not found error because the URI is not
+        # contained in the CRISPR test data
+        uri = 'http://sbols.org/CRISPR_Example/mKate_genie/1.0.0'
+        with self.assertRaises(sbol2.SBOLError) as cm:
+            doc.getExtensionObject(uri)
+        raised = cm.exception
+        self.assertEqual(sbol2.SBOLErrorCode.SBOL_ERROR_NOT_FOUND,
+                         raised.error_code())
+
+    def test_add_extension_object(self):
+        # We need to construct and add an extension object
+        # It should then be accessible via getExtensionObject
+        # We can also peel back the curtain and make sure the object
+        #  is in doc.SBOLObjects and the URI is not in doc.OwnedObjects
+        # Three kinds of things to add:
+        #  1. An existing type, like ComponentDefinition
+        cd = sbol2.ComponentDefinition('cd')
+        doc = sbol2.Document()
+        doc.addExtensionObject(cd)
+        obj = doc.getExtensionObject(cd.identity)
```

Comment:
At this line, what would be the result of...?
```
self.assertEqual(len(doc.componentDefinitions), 1)
```
There might be some janky behavior here, as a result of adding a Core object as an Extension object.

Reply:
The assertEqual passes. `len(doc.componentDefinitions) == 1`

Reply:
OK, this is the ideal behavior.  I suggest including that assertion in the test, if there's not already a similar test for the underlying `Document.add` method.  I'm not at all sure pySBOL would actually pass this test.

Reply:
I added that assert to the unit test.